File: 0207-course-schedule/0207-course-schedule.ts
function canFinish(numCourses: number, prerequisites: number[][]): boolean {
    const graph: Record<string, number[]> = {}, indegree: Record<string, number> = {}, q: number[] = []
    let count = 0
    for (let n=0; n < numCourses; n++) {
        indegree[n] = 0
        graph[n] = []
    }
    for (const [course, pre] of prerequisites) {
        indegree[course]++
        graph[pre].push(course)
    }
    for (let n=0; n < numCourses; n++) {
        if (indegree[n] === 0) {
            q.push(n)
        }
    }
    while (q.length) {
        const i = q.shift()
        count++
        for (const j of graph[i] || []) {
            indegree[j]--
            if (indegree[j] === 0) {
                q.push(j)
            }
        }
    }
    return count === numCourses
};

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        in_degree, graph, q, count = {}, {}, deque([]), 0
        for n in range(0, numCourses):
            in_degree[n] = 0
            graph[n] = []
        for course, prereq in prerequisites:
            in_degree[course] += 1
            graph[prereq].append(course)
        for n in range(0, numCourses):
            if in_degree[n] == 0:
                q.append(n)
        while q:
            n = q.popleft()
            count += 1
            for next in graph[n]:
                in_degree[next] -= 1
                if in_degree[next] == 0:
                    q.append(next)
        return count == numCourses
            